src/mame/mattel/pixtermu.cpp

```diff
@@ -0,0 +1,223 @@
+{
+public:
+	pixter_multimedia_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag)
+		, m_cart(*this, "cartslot")
+		, m_maincpu(*this, "maincpu")
+		, m_ndcs0(*this, "ndcs0")
+		, m_internal_sram(*this, "internal_sram")
+		, m_apb(*this, "apb", 0x27000, ENDIANNESS_LITTLE)
+		, m_remap_view(*this, "remap")
+	{ }
+
+	void pixter_multimedia(machine_config &config);
+
+private:
+	// Remap Control, mapped at 0xfffe2008, offset 0x22008/4
+	static inline constexpr uint32_t APB_REMAP = 34818;
+	// Power-up Boot Configuration, mapped at 0xfffe6000, offset 0x26000/4
+	static inline constexpr uint32_t APB_PBC = 38912;
+	// nCS1 Override, mapped at 0xfffe6004, offset 0x26004/4
+	static inline constexpr uint32_t APB_CS1OV = 38913;
+	// External Peripheral Mapping, mapped at 0xfffe6008, offset 0x26008/4
+	static inline constexpr uint32_t APB_EPM = 38914;
+
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+	DEVICE_IMAGE_LOAD_MEMBER(cart_load);
+	void arm7_map(address_map &map);
+	void apb_bridge_w(offs_t offset, uint32_t data, uint32_t mem_mask);
+	void apb_remap(uint32_t data);
+
+	required_device<generic_slot_device> m_cart;
+	required_device<arm7_cpu_device> m_maincpu;
+	required_shared_ptr<uint32_t> m_ndcs0;
+	required_shared_ptr<uint32_t> m_internal_sram;
+	memory_share_creator<uint32_t> m_apb;
+	memory_view m_remap_view;
+};
+
+void pixter_multimedia_state::apb_remap(uint32_t data)
+{
+	// User's Guide - 1.6 Memory Interface Architecture
+	if (data == 0) {
+		m_remap_view.select((m_apb[APB_PBC] & 0b0100) && (m_apb[APB_CS1OV] & 0b1) ? 0 : 3);
+	} else if (data < 3) {
+		m_remap_view.select(data);
+	} else {
+		logerror("Unexpected remap %d\n", data);
+	}
+}
+
+void pixter_multimedia_state::machine_start()
+{
+	if (m_cart->exists()) {
+		memory_region *cart_rom = m_cart->memregion("cart:rom");
+		device_generic_cart_interface::install_non_power_of_two<0>(
+			cart_rom->bytes(),
+			0x03ffffff,
+			0,
+			0x48000000,
+			[this, cart_rom](offs_t begin, offs_t end, offs_t mirror, offs_t src) {
+				m_maincpu->space(AS_PROGRAM).install_rom(begin, end, mirror, cart_rom->base() + src);
+			});
+	}
+}
+
+void pixter_multimedia_state::machine_reset()
+{
+	m_apb[APB_PBC] = 0b0000; // Boot from NOR Flash or SRAM, 16-bit data bus width, nBLEx LOW for reads
+	m_apb[APB_CS1OV] = 0b0; // nCS1 is routed for normal operation
+	m_apb[APB_EPM] = 0b1111; // All external devices are accessible following reset
+
+	m_apb[APB_REMAP] = 0b00; // Map nCS1
+	apb_remap(m_apb[APB_REMAP]);
+}
+
+DEVICE_IMAGE_LOAD_MEMBER(pixter_multimedia_state::cart_load)
+{
+	if (!image.loaded_through_softlist()) {
+	   uint32_t const size = m_cart->common_get_size("rom");
+
+	   m_cart->rom_alloc(size, GENERIC_ROM32_WIDTH, ENDIANNESS_LITTLE);
+	   m_cart->common_load_rom(m_cart->get_rom_base(), size, "rom");
+	}
+
+	return std::make_pair(std::error_condition(), std::string());
+}
```

Comment:
OK, we’re almost there, but there are still a few issues:
* No error checking to ensure that the software list entry is usable (people can edit software lists, and easily screw things up)
* `rom_alloc` won’t give the same tag as the region loaded from the software list
* `common_load_rom` doesn’t report errors
* You need to remember to swap the data on big Endian hosts

Something like this should work, but I haven’t tested it, just coding from memory:
```c++
DEVICE_IMAGE_LOAD_MEMBER(pixter_multimedia_state::cart_load)
{
	uint64_t length;
	memory_region *cart_rom = nullptr;
	if (m_cart->loaded_through_softlist()) {
		cart_rom = m_cart->memregion("rom");
		if (!cart_rom)
			return std::make_pair(image_error::BADSOFTWARE, "Software list item has no 'rom' data area");
		length = cart_rom->bytes();
	} else {
		length = m_cart->length();
	}

	if (!length)
		return std::make_pair(image_error::INVALIDLENGTH, "Cartridges must not be empty");
	if (length & 3)
		return std::make_pair(image_error::INVALIDLENGTH, "Unsupported cartridge size (must be a multiple of 4 bytes)");

	if (!m_cart->loaded_through_softlist()) {
		cart_rom = machine().memory().region_alloc(m_cart->subtag("rom"), length, 4, ENDIANNESS_LITTLE);
		if (!cart_rom)
			return std::make_pair(std::errc::not_enough_memory, std::string());

		uint32_t *const base = reinterpret_cast<uint32_t *>(cart_rom->base());
		if (m_cart->fread(base, length) != length)
			return std::make_pair(std::errc::io_error, "Error reading cartridge file");

		if (ENDIANNESS_NATIVE != ENDIANNESS_LITTLE)
		{
			for (uint64_t i = 0; (length / 4) > i; ++i)
				base[i] = swapendian_int32(base[i]);
		}
	}

	return std::make_pair(std::error_condition(), std::string());
}
```

Reply:
Yes, I realise it would be useful to have a helper function to implement this pattern or something similar rather than the problematic `common_load_rom`.  Replacing the current bus/generic stuff is a long-term goal.

Reply:
Works fine, thanks.